notation.go

```diff
@@ -73,48 +58,57 @@ type signerAnnotation interface {
 	PluginAnnotations() map[string]string
 }
 
-// Sign signs the artifact in the remote registry and push the signature to the
-// remote.
+// SignOptions contains parameters for notation.Sign.
+type SignOptions struct {
+	SignerSignOptions
+
+	// ArtifactReference sets the reference of the artifact that needs to be
+	// signed. It can be a tag, a digest or a full reference.
+	ArtifactReference string
+
+	// UserMetadata contains key-value pairs that are added to the signature
+	// payload
+	UserMetadata map[string]string
+}
+
+// Sign signs the artifact and push the signature to the Repository.
 // The descriptor of the sign content is returned upon sucessful signing.
-func Sign(ctx context.Context, signer Signer, repo registry.Repository, remoteOpts RemoteSignOptions) (ocispec.Descriptor, error) {
-	// Input validation for expiry duration
-	if remoteOpts.ExpiryDuration < 0 {
+func Sign(ctx context.Context, signer Signer, repo registry.Repository, signOpts SignOptions) (ocispec.Descriptor, error) {
+	// sanity check
+	if signer == nil {
+		return ocispec.Descriptor{}, errors.New("signer cannot be nil")
+	}
+	if repo == nil {
+		return ocispec.Descriptor{}, errors.New("repo cannot be nil")
+	}
+	if signOpts.ExpiryDuration < 0 {
 		return ocispec.Descriptor{}, fmt.Errorf("expiry duration cannot be a negative value")
 	}
-
-	if remoteOpts.ExpiryDuration%time.Second != 0 {
+	if signOpts.ExpiryDuration%time.Second != 0 {
 		return ocispec.Descriptor{}, fmt.Errorf("expiry duration supports minimum granularity of seconds")
 	}
 
 	logger := log.GetLogger(ctx)
-	artifactRef := remoteOpts.ArtifactReference
-	ref, err := orasRegistry.ParseReference(artifactRef)
+	targetDesc, err := repo.Resolve(ctx, signOpts.ArtifactReference)
```

Comment:
There is an edge case that `ArtifactReference` is a full reference but `repo` can only resolve a tag or a digest.

Reply:
> There is an edge case that `ArtifactReference` is a full reference but `repo` can only resolve a tag or a digest.

This happens when `repo` is an `oci.Store`. In this case, if `ArtifactReference` is a full reference, `targetDesc, err := repo.Resolve(ctx, signOpts.ArtifactReference)` would return an error.

Reply:
Updated the code after discussion.